### ISSUE TITLE
ERROR: setindex! not defined for Zygote.OneElement{...}

### ISSUE THREAD

**Issue by mkschleg** [FluxML/Flux.jl#1626]
Working with RNNs. Found a weird issue when updating to 0.12.4, but might be more related to an update with zygote (hard to tell). Basically, if I don't pass in features to the RNN as a vector I get `ERROR: setindex! not defined for Zygote.OneElement{...}` when trying to update the parameters. Seems like this just needs to be implemented as the gradient (which is of type `Zygote.OneElement{...}` looks vectorish). 

If you instead use `x=[rand(Float32, 5, 1) for i in 1:10];` below, you don't get the error.

```julia
julia> m = Chain(GRU(5, 10), Dense(10, 3))
Chain(Recur(GRUCell(5, 10)), Dense(10, 3))

julia> x = [rand(Float32, 5) for i in 1:10];

julia> a_t = 2

julia> loss(p, a_t, target) = (p[a_t] - target)^2
loss (generic function with 2 methods)

julia> opt = RMSProp()
RMSProp(0.001, 0.9, IdDict{Any, Any}())

julia> gs = gradient(Flux.params(m)) do
           preds = map(m, x)
           target = Flux.Zygote.dropgrad(maximum(preds[end]))
           loss(preds[end-1], a_t, target)
       end
Grads(...)

julia> Flux.Optimise.update!(opt, Flux.params(m), gs)
ERROR: setindex! not defined for Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}}
Stacktrace:
  [1] error(::String, ::Type)
    @ Base ./error.jl:42
  [2] error_if_canonical_setindex(#unused#::IndexCartesian, A::Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}}, #unused#::Int64)
    @ Base ./abstractarray.jl:1277
  [3] setindex!
    @ ./abstractarray.jl:1266 [inlined]
  [4] macro expansion
    @ ./broadcast.jl:984 [inlined]
  [5] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [6] copyto!
    @ ./broadcast.jl:983 [inlined]
  [7] copyto!
    @ ./broadcast.jl:936 [inlined]
  [8] materialize!
    @ ./broadcast.jl:894 [inlined]
  [9] materialize!
    @ ./broadcast.jl:891 [inlined]
 [10] apply!(o::RMSProp, x::Vector{Float32}, Δ::Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}})
    @ Flux.Optimise ~/.julia/packages/Flux/0c9kI/src/optimise/optimisers.jl:145
 [11] update!(opt::RMSProp, x::Vector{Float32}, x̄::Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}})
    @ Flux.Optimise ~/.julia/packages/Flux/0c9kI/src/optimise/train.jl:23
 [12] update!(opt::RMSProp, xs::Zygote.Params, gs::Zygote.Grads)
    @ Flux.Optimise ~/.julia/packages/Flux/0c9kI/src/optimise/train.jl:29
 [13] top-level scope
    @ REPL[34]:1
```

**Comment by mkschleg** [comment]
btw:
```julia
(ActionRNNs) pkg> st -m Flux Zygote
     Project ActionRNNs v0.1.0
      Status `~/Documents/Research/projects/ActionRNNs/Manifest.toml`
  [587475ba] Flux v0.12.4
  [e88e6eb3] Zygote v0.6.14
```

**Comment by darsnack** [comment]
This is the same issue as https://github.com/FluxML/Flux.jl/issues/1592 but for `OneElement` instead of views.

**Comment by mkschleg** [comment]
Cool! I'm not sure I'm following that thread well. Is there a fix there? Or am I missing something?

**Comment by darsnack** [comment]
The issue is the same meaning that certain optimizers like `RMSProp` or `ADAM` mutate the gradient in-place. In these cases we need to be more careful about the types that we use.

But I see after a more careful inspection that your particular issue is resolved on master by #1613. Could you test with `]add Flux#master`? If it works, we can close.

**Comment by mkschleg** [comment]
That makes sense. But maybe I'm still being a bit dense, as this also happens when using just SGD:

```julia
julia> opt = Descent();

julia> gs = gradient(Flux.params(m)) do
           preds = map(m, x)
           target = Flux.Zygote.dropgrad(maximum(preds[end]))
           loss(preds[end-1], a_t, target)
       end
Grads(...)

julia> Flux.Optimise.update!(opt, Flux.params(m), gs)
ERROR: setindex! not defined for Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}}
Stacktrace:
  [1] error(::String, ::Type)
    @ Base ./error.jl:42
  [2] error_if_canonical_setindex(#unused#::IndexCartesian, A::Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}}, #unused#::Int64)
    @ Base ./abstractarray.jl:1277
  [3] setindex!
    @ ./abstractarray.jl:1266 [inlined]
  [4] macro expansion
    @ ./broadcast.jl:984 [inlined]
  [5] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [6] copyto!
    @ ./broadcast.jl:983 [inlined]
  [7] copyto!
    @ ./broadcast.jl:936 [inlined]
  [8] materialize!
    @ ./broadcast.jl:894 [inlined]
  [9] materialize!
    @ ./broadcast.jl:891 [inlined]
 [10] apply!
    @ ~/.julia/packages/Flux/0c9kI/src/optimise/optimisers.jl:42 [inlined]
 [11] update!(opt::Descent, x::Vector{Float32}, x̄::Zygote.OneElement{Float32, 1, Tuple{Int64}, Tuple{Base.OneTo{Int64}}})
    @ Flux.Optimise ~/.julia/packages/Flux/0c9kI/src/optimise/train.jl:23
 [12] update!(opt::Descent, xs::Zygote.Params, gs::Zygote.Grads)
    @ Flux.Optimise ~/.julia/packages/Flux/0c9kI/src/optimise/train.jl:29
 [13] top-level scope
    @ REPL[43]:1
```

**Comment by mkschleg** [comment]
I'll try master though.

**Comment by mkschleg** [comment]
I was being dense. Works on master. Will close.

**Comment by darsnack** [comment]
No, that's a mistake on my part. All the optimizers mutate the gradient in-place, so this is an issue no matter which one you use. #1592 is the same underlying issue but for the momentum terms (which is why I mentioned `RMSProp`). Either way, it should be resolved by #1613.
